### PR TITLE
NV-2159 - 🚀 Feature: Uncancel cancelled events

### PR DESCRIPTION
**Pull request by gitstart**
### What change does this PR introduce?
 This feature enables the user to resume a canceled job if it was meant to be sent afterall.

 ### Why was this change needed?
closes https://github.com/novuhq/novu/issues/3270

### Other information

---
This code was written and reviewed by GitStart Community. Growing great engineers, one PR at a time.

### CODE REVIEW

**Comment by gitstart** [comment]
@scopsy this PR is ready for review

**Review by p-fernandez — CHANGES_REQUESTED** [review]

**Comment by gitstart** [comment]
@p-fernandez your comments have been addressed, thanks for the feedback.

File: apps/api/src/app/events/e2e/delay-events.e2e.ts [hunk 1]
```diff
@@ -12,8 +12,10 @@ import {
 import { UserSession, SubscribersService } from '@novu/testing';
 import { StepTypeEnum, DelayTypeEnum, DigestUnitEnum, DigestTypeEnum } from '@novu/shared';
 import { QueueService } from '@novu/application-generic';
+import { setTimeout } from 'timers/promises';
 
 const axiosInstance = axios.create();
+const promiseResolveTimeout = (ms: number) => setTimeout(ms);
```
Comment (p-fernandez) [inline apps/api/src/app/events/e2e/delay-events.e2e.ts:18]: This is not needed. Just call `await setTimeout` directly wherever it is needed.

Comment (gitstart) [reply apps/api/src/app/events/e2e/delay-events.e2e.ts]: Done @p-fernandez

**Review by p-fernandez — APPROVED** [review]

File: apps/api/src/app/events/events.controller.ts [hunk 2]
```diff
@@ -21,8 +21,11 @@ import { TriggerEventToAll, TriggerEventToAllCommand } from './usecases/trigger-
 import { UserSession } from '../shared/framework/user.decorator';
 import { ExternalApiAccessible } from '../auth/framework/external-api.decorator';
 import { JwtAuthGuard } from '../auth/framework/auth.guard';
+import { ResumeDelayedCommand } from './usecases/resume-delayed';
```
Comment (p-fernandez) [inline apps/api/src/app/events/events.controller.ts:24]:
```suggestion
import { ResumeDelayed, ResumeDelayedCommand } from './usecases/resume-delayed';
```

File: apps/api/src/app/events/events.controller.ts [hunk 3]
```diff
@@ -21,8 +21,11 @@ import { TriggerEventToAll, TriggerEventToAllCommand } from './usecases/trigger-
...
 import { ApiResponse } from '../shared/framework/response.decorator';
 import { DataBooleanDto } from '../shared/dtos/data-wrapper-dto';
+import { ResumeDelayed } from 'apps/api/dist/src/app/events/usecases/resume-delayed';
```
Comment (p-fernandez) [inline apps/api/src/app/events/events.controller.ts:27]:
```suggestion
```

**Review by p-fernandez — CHANGES_REQUESTED** [review]
These both changes are needed so tests will work. This wrongly linked to the compiled code.

File: apps/api/src/app/events/events.controller.ts [hunk 4]
```diff
@@ -177,6 +180,28 @@ export class EventsController {
...
 
+  @ExternalApiAccessible()
+  @UseGuards(JwtAuthGuard)
+  @Post('/trigger/:transactionId/resume')
+  @ApiNoContentResponse()
+  @ApiOperation({
+    summary: 'Resume canceled event',
+    description: ` 
+      Using this endpoint, you can resume canceled workflows that are yet to expire.
+      This is especially useful when a canceled message, is meant to be sent afterall.
```
Comment (p-fernandez) [inline apps/api/src/app/events/events.controller.ts:191]:
```suggestion
      This is especially useful when a canceled message, is meant to be sent after all.
```

**Review by p-fernandez — APPROVED** [review]
Now I think will be all good. Thank you for the contribution helping us to close this feature.
🌟

**Review by p-fernandez — CHANGES_REQUESTED** [review]
I was too positive. These tests need to be checked:
Trigger event - Delay triggered events - /v1/events/trigger (POST)
Trigger event - Delay triggered events - /v1/events/trigger (POST)
Probably they were dependant on a previous state of another test. 😢

**Comment by gitstart** [comment]
@p-fernandez the tests have been duly updated, thanks for the feedback.

File: apps/api/src/app/events/usecases/resume-delayed/resume-delayed.usecase.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+      _environmentId: command.environmentId,
+      transactionId: command.transactionId,
+      status: JobStatusEnum.CANCELED,
+    });
+
+    if (!job) {
+      throw new NotFoundException(`Job with transactionId ${command.transactionId} was not found`);
+    }
+
+    if (isBefore(new Date(job?.expireAt as string), new Date())) {
```
Comment (ainouzgali) [inline apps/api/src/app/events/usecases/resume-delayed/resume-delayed.usecase.ts:22]: expireAt is the ttl value, its the date when the entity will be archived. It is unrelated to the delay amount. 
You need to check if the calculation of the delay did not pass

File: apps/api/src/app/events/events.controller.ts (hunk 4 above)
Comment (ainouzgali) [inline apps/api/src/app/events/events.controller.ts:191]:
```suggestion
      This is especially useful when a canceled message, is meant to be sent after all.
```

File: apps/api/src/app/events/e2e/delay-events.e2e.ts
Comment (ainouzgali) [inline apps/api/src/app/events/e2e/delay-events.e2e.ts:1]: You should please add another test that checks that after resuming -> the jobs after that are processed. Like - check that an email message after the delay step has been sent

**Comment by gitstart** [comment]
@ainouzgali thanks for the review. Will update the PR with the fixes soon!

**Comment by scopsy** [comment]
Closing as stale 🙏

**Pull request closed without merging**